rust/worker/src/execution/operators/limit.rs

```diff
@@ -0,0 +1,317 @@
+}
+
+#[derive(Error, Debug)]
+pub enum LimitError {
+    #[error("Error creating record segment reader {0}")]
+    RecordSegmentReaderCreationError(#[from] RecordSegmentReaderCreationError),
+    #[error("Error materializing logs {0}")]
+    LogMaterializationError(#[from] LogMaterializerError),
+    #[error("Error reading from record segment")]
+    RecordSegmentReaderError,
+}
+
+impl ChromaError for LimitError {
+    fn code(&self) -> ErrorCodes {
+        match self {
+            LimitError::RecordSegmentReaderCreationError(e) => e.code(),
+            LimitError::LogMaterializationError(e) => e.code(),
+            LimitError::RecordSegmentReaderError => ErrorCodes::Internal,
+        }
+    }
+}
+
+// This struct aims to help scanning a number of elements from a fixed offset in the imaginary
+// segment where the log is compacted and the element in the mask is ignored.
+struct SkipScanner<'me> {
+    log_oids: &'me RoaringBitmap,
+    record_segment: &'me RecordSegmentReader<'me>,
+    mask: &'me RoaringBitmap,
+}
+
+impl<'me> SkipScanner<'me> {
+    // Find the rank of the target offset id in the imaginary segment
+    async fn joint_rank(&self, target: u32) -> Result<usize, LimitError> {
+        // # of elements strictly less than target in materialized log
+        let log_rank =
+            self.log_oids.rank(target) as usize - self.log_oids.contains(target) as usize;
+        // # of elements strictly less than target in the record segment
+        let record_rank = self
+            .record_segment
+            .get_offset_id_rank(target)
+            .await
+            .map_err(|_| LimitError::RecordSegmentReaderError)?;
+        // # of elements strictly less than target in the mask
+        let mask_rank = self.mask.rank(target) as usize - self.mask.contains(target) as usize;
+        Ok(log_rank + record_rank - mask_rank)
+    }
+
+    // Skip to the start in log and record segment
+    // The following implementation is based on std implementation
+    async fn skip_to_start(&self, skip: usize) -> Result<(usize, usize), LimitError> {
+        if skip == 0 {
+            return Ok((0, 0));
+        }
+
+        let mut size = self
+            .record_segment
+            .get_current_max_offset_id()
+            .load(atomic::Ordering::Relaxed)
+            .max(self.log_oids.max().unwrap_or(0));
+        if size == 0 {
+            return Ok((0, 0));
+        }
+
+        let mut base = 0;
+        while size > 1 {
+            let half = size / 2;
+            let mid = base + half;
+
+            let cmp = self.joint_rank(mid).await?.cmp(&skip);
+            base = if cmp == Ordering::Greater { base } else { mid };
+            size -= half;
+        }
+
+        Ok((
+            self.log_oids.rank(base) as usize - self.log_oids.contains(base) as usize,
+            self.record_segment
+                .get_offset_id_rank(base)
+                .await
+                .map_err(|_| LimitError::RecordSegmentReaderError)?,
```

Comment:
nit: may be more readable to not include the record segment logic in this function since it's just a single method call and the majority of logic here is dealing with the log index

Reply:
Refactored in a later PR.